Combined frontend and backend for dashboard and creating event

ZhenyuYu1 opened the pull request:
Features added:
1. Combined frontend UI with backend
2. Dashboard pulls events from events table from database
3. Submitting an event through the events form inserts an event into the events table
4. API routes for getting and posting event

ZhenyuYu1 committed:
added get/post for users and events table, seed data for tables

ZhenyuYu1 committed:
added get/post for users and events table, seed data for tables

ZhenyuYu1 committed:
Merge branch 'main' of https://github.com/ZhenyuYu1/SchedulerClone into dbSetup

ZhenyuYu1 committed:
added input validation for events

AlanaDubie committed:
added /create-event page and EventForm.tsx component for Create Event Dashboard view (left hand side). Removed .dark styling class in globals.css

AlanaDubie committed:
added times function to generate hours from 1AM to 12PM for earliestTimes and latestTimes. Added handleSelectedDayOfWeek function to handle selected daysOfWeek array based on checkbox selection and deselection. Added values for timezone dropdown

ZhenyuYu1 committed:
added api to retrieve and insert event data, added local storage for username

Jean-LucMartel committed:
updating homepage

ZhenyuYu1 committed:
Merge branch 'jm-event-homepage-2' of https://github.com/ZhenyuYu1/SchedulerClone into combine-frontend-backend

ZhenyuYu1 committed:
Merge branch 'event-form-dashboard' of https://github.com/ZhenyuYu1/SchedulerClone into combine-frontend-backend

ZhenyuYu1 committed:
combined backend with frontend for dashboard and creating event

ZhenyuYu1 committed:
removed dashboard file

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+
+  if (body.title.length > 120) {
+    return NextResponse.json(
+      { message: 'Title must be less than 120 characters' },
```
Small nitpick - this should be ">= 120" or "Title must be less than or equal to 120 characters"

ZhenyuYu1 replied:
Agreed, changed to "Title must be less than or equal to 120 characters".

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+  } else if (body.description && body.description.length > 500) {
+    // description is optional
+    return NextResponse.json(
+      { message: 'Description must be less than 500 characters' },
```
Same as above.

ZhenyuYu1 replied:
Agreed, changed to 'Description must be less than or equal to 500 characters'.

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+      { message: 'Description must be less than 500 characters' },
+      { status: 400 },
+    )
+  } else if (body.location.length > 120) {
```
Same as above.

ZhenyuYu1 replied:
Agreed, changed to 'Location must be less than or equal to 120 characters'.

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+      { message: 'Location must be less than 120 characters' },
+      { status: 400 },
+    )
+  } else if (body.mode != 'weekly' && body.mode != 'specific') {
```
This might be better of as an enum or an array of valid values that can be shared.  A common data model makes it usable in other parts of the code base.

array example:
const modeOptions = ['weekly', 'specific'];
if ( modeOptions.includes(body.mode) ) {

ZhenyuYu1 replied:
Agreed.

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+    )
+  } else if (body.mode != 'weekly' && body.mode != 'specific') {
+    return NextResponse.json(
+      { message: 'Mode must be weekly or specific' },
```
If you use the array you could programatically print the array instead.

ZhenyuYu1 replied:
Agreed.

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+
+  // if mode is weekly, check if each day is a true/false boolean
+  if (body.mode == 'weekly') {
+    const days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']
```
This might be better placed in a common file that can be used by others.

ZhenyuYu1 replied:
Agreed, added a dateUtils.ts file with the days array and modeOptions enum.

pselvana commented on src/app/api/events/create/route.ts:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  // if mode is specific, check if dates in config are valid dates in format (YYYY-MM-DD)
+  if (body.mode == 'specific') {
```
to be type safe ,should:
body.mode === 'specific'

ZhenyuYu1 replied:
Agreed.

pselvana commented on src/app/api/users/create/route.ts:
```diff
@@ -0,0 +1,35 @@
+    )
+  } else if (body.name.length > 40) {
+    return NextResponse.json(
+      { message: 'Name must be less than 40 characters' },
```
same as events create; less than 40 vs ">40"

ZhenyuYu1 replied:
Agreed.

pselvana commented on src/app/create-event/page.tsx:
```diff
@@ -0,0 +1,121 @@
+      daysOfWeek.length > 0
+        ? daysOfWeekJSON
+        : JSON.parse('{"days": ["2024-01-01"]}'),
+      '9e33186f-95db-4385-a974-ee38c8e07547',
```
What's this for?

ZhenyuYu1 replied:
Changed to mode === 'weekly' ? daysOfWeekJSON : JSON.parse('{"days": ["2024-01-01"]}').

pselvana commented on src/app/create-event/page.tsx:
```diff
@@ -0,0 +1,121 @@
+      daysOfWeek.length > 0 ? 'weekly' : 'specific',
+      daysOfWeek.length > 0
+        ? daysOfWeekJSON
+        : JSON.parse('{"days": ["2024-01-01"]}'),
```
Is this a default?

ZhenyuYu1 replied:
Yes, the {"days": ...} is filler since we don't have a way to get specific days yet with a calendar.

pselvana commented on src/app/dashboard/page.tsx:
```diff
@@ -0,0 +1,221 @@
+      }
+    }
+
+    getMyEvents('9e33186f-95db-4385-a974-ee38c8e07547')
```
It's back! What's this for?

ZhenyuYu1 replied:
Filler UUID until local storage to store a user ID on event creation or visiting an event page is setup so that we can use the user's id in local storage to get events isntead. I have added this specific UUID as an id for a user in the seed.sql so that it should be work with local databases.

pselvana commented on src/app/dashboard/page.tsx:
```diff
@@ -0,0 +1,221 @@
+      JSON.parse(
+        '{"days": ["2024-07-17", "2024-07-18", "2024-07-19", "2024-07-20", "2024-07-21", "2024-07-22", "2024-07-23"]}',
+      ),
+      '9e33186f-95db-4385-a974-ee38c8e07547', // random UUID,
```
if you need a random UUID you can use something like this: https://supabase.com/docs/guides/database/extensions/uuid-ossp

pselvana replied:
Scratch that... looks like you can do something like what you have below for in the schema and let default carry the work:
id uuid primary key default gen_random_uuid()

ZhenyuYu1 replied:
I think this will be resolved once I setup the local storage to store the user's id since we can then use that id when generating an event. Right now this is a filler id that is also in the seed.sql for local development purposes.

pselvana commented on src/app/globals.css:
```diff
@@ -25,28 +25,6 @@
     --ring: 240 5.9% 10%;
     --radius: 0.3rem;
   }
-
```
Can you remove this change please? I think I have a fix to get rid of this problem that is in the Starter code.

pselvana commented on src/app/page.tsx:
```diff
@@ -1,57 +1,90 @@
+        })
+        .then((data) => {
+          const dataFiltered = data.filter(
+            (event: any) => event.creator === creatorId,
```
Please move this to the GET call by passing in a query parameter of the local user ID.  

Something like this might work:
export async function GET(request: Request, { params }: { params: { creator: string } }) {
  const { creator } = params;

ZhenyuYu1 replied:
Agreed. Done using query parameters in the URL.

pselvana commented on src/components/EventCard.tsx:
```diff
@@ -0,0 +1,21 @@
+  endTime: string
+}
+
+export default function EventCard({
```
Great to see this refactored into a component.

pselvana commented on src/components/EventCard.tsx:
```diff
@@ -0,0 +1,21 @@
+  endTime,
+}: EventCardProps) {
+  return (
+    <div className="flex h-48 flex-col justify-between rounded-md bg-white p-4 shadow-lg">
```
I personally like https://daisyui.com/components/card/ but I'll leave it you if you prefer this component.

pselvana replied:
As discussed, Jean-Luc wants to keep what is here.

pselvana commented on src/components/EventForm.tsx:
```diff
@@ -0,0 +1,222 @@
+              type="checkbox"
+              name="options"
+              aria-label="Mon"
+              checked={daysOfWeek.includes('Mon')}
```
This should be refactored.
Maybe something like:


const days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];

const handleChange = (day) => {
  handleSelectedDayOfWeek(day);
};

{days.map(day => (
  <input
    key={day}
    className="btn join-item border-gray-300 bg-white"
    type="checkbox"
    name="options"
    aria-label={day}
    checked={daysOfWeek.includes(day)}
    onChange={() => handleChange(day)}
  />
))}

pselvana replied:
@AlanaDubie  - Can you please complete this in a follow-up change? the "const days" is available in a file Zhenyu created.  Closing this for now.

pselvana commented:
Team - let me know if the comments resonate. If you agree, make the change and test and commit back to this branch. If you think you want to do a follow up change, create a card on the "Backlog" in Trello and prefix the word "[BUG]" to the title so we can track getting those fixed up.  If you think you don't want to make the change, please justify why.  If you have questions, please comment back.

ZhenyuYu1 committed:
fixed wordings for character limit restrictions, refactor code to use mode to
check for correct config, changed seed.sql to put in filler UUIDs for testing purposes

ZhenyuYu1 committed:
Merge branch 'main' of https://github.com/ZhenyuYu1/SchedulerClone into combine-frontend-backend

ZhenyuYu1 committed:
fixing bug with prettier

ZhenyuYu1 committed:
revert global.css change

ZhenyuYu1 committed:
renamed loc to location and descript to description in seed.sql for events table

ZhenyuYu1 committed:
changed events table to use all lowercase column names

ZhenyuYu1 committed:
changed startTime and endTime to starttime and endtime in create events

ZhenyuYu1 committed:
rename startTime and endTime to stattime and endtime in eventCard

pselvana committed:
Fix event insertion and read problem
